lib/migration/migrator.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/vmware/vic/lib/migration/errors"
+	"github.com/vmware/vic/lib/migration/manager"
+	_ "github.com/vmware/vic/lib/migration/plugins"
+	"github.com/vmware/vic/pkg/vsphere/session"
+)
+
+// MigrateApplianceConfigure migrate VCH appliance configuration, including guestinfo, keyvaluestore, or any other kinds of change
+// InIt accept VCH appliance guestinfo map, and return all configurations need to be made in guestinfo, keyvaluestore, and can have
+// more kinds of change in the future. Each kind is one map key/value pair.
+// If there is error returned, returned map might have half-migrated value, this is why we don't persist any data in plugin.
+//func MigrateApplianceConfigure(data map[string]string) (map[string]string, bool, error) {
+func MigrateApplianceConfigure(ctx context.Context, s *session.Session, conf map[string]string) (map[string]string, bool, error) {
+	return migrateConfigure(ctx, s, conf, manager.ApplianceConfigure, manager.ApplianceVersionKey)
+}
+
+// MigrateContainerConfigure migrate container configuration
+// Migrated data will be returned in map, and input object is not changed.
+// If there is error returned, returned map might have half-migrated value.
+func MigrateContainerConfigure(conf map[string]string) (map[string]string, bool, error) {
+	return migrateConfigure(nil, nil, conf, manager.ContainerConfigure, manager.ContainerVersionKey)
+}
+
+func IsContainerDataOlder(conf map[string]string) (bool, error) {
+	return isDataOlder(conf, manager.ContainerConfigure, manager.ContainerVersionKey)
+}
+
+func IsApplianceDataOlder(conf map[string]string) (bool, error) {
+	return isDataOlder(conf, manager.ApplianceConfigure, manager.ApplianceVersionKey)
+}
+
+func isDataOlder(data map[string]string, target string, verKey string) (bool, error) {
+	var currentID int
+	var err error
+
+	if currentID, err = getCurrentID(data, verKey); err != nil {
+		return false, err
+	}
+	latestVer := manager.Migrator.GetLatestVersion(target)
+	return latestVer > currentID, nil
+}
+
```

Comment:
Multiple return values need comment, otherwise its hard to understand. At this point I have troubles to get an idea what that bool is.

migrateConfigure -> migrateConfig or migrateConfiguration.
Configure is a verb.